### PR TITLE
integrated initial draft of parser plugin writing

### PR DESCRIPTION
**Pull request by JFRudzinski**
Just added my preliminary version. I will continue to work on this one after I finish with my H5MD stuff.

### CODE REVIEW

File: docs/writing_a_parser_plugin/computational.md [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+     ├── tasks
+     ├── outputs
+     └── results
+```
+
+The most important section of the archive for computational data is the `run` section, which is
+divided into three main subsections: `method`, `system`, and `calculation`. `method` stores
+information about the computational model used to perform the calculation. `system` stores
+attributes of the atoms involved in the calculation, e.g., atom types, positions, lattice vectors, etc.
+`calculation` stores the output of the calculation, e.g., energy, forces, etc.
```
Comment (ndaelman-hu) [inline docs/writing_a_parser_plugin/computational.md]: I would highlight the semantics of each section, since this is the main information to be communicated quickly. e.g.

> The most important section of the archive for **computational data** is the `run` section

File: docs/writing_a_parser_plugin/computational.md [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+
+The `workflow` section of the archive then stores information about the series of tasks performed
+to accumulate the (meta)data in the run section. The relevant input parameters for the workflow are
+stored in `method`, while the `results` section stores output from the workflow beyond observables
+of single configurations. For example, any ensemble-averaged quantity from a molecular dynamics
```
Comment (ndaelman-hu) [inline docs/writing_a_parser_plugin/computational.md]: Wouldn't it be easier to say that its subsection reference other sections in `run`?
I think this better summarizes the general rule.

File: docs/writing_a_parser_plugin/computational.md [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+     └── results
+```
+
+The most important section of the archive for computational data is the `run` section, which is
+divided into three main subsections: `method`, `system`, and `calculation`. `method` stores
+information about the computational model used to perform the calculation. `system` stores
+attributes of the atoms involved in the calculation, e.g., atom types, positions, lattice vectors, etc.
+`calculation` stores the output of the calculation, e.g., energy, forces, etc.
+
+The `workflow` section of the archive then stores information about the series of tasks performed
```
Comment (ndaelman-hu) [inline docs/writing_a_parser_plugin/computational.md:45]: I think it is still called `workflow2` for now.

File: docs/writing_a_parser_plugin/computational.md [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+to accumulate the (meta)data in the run section. The relevant input parameters for the workflow are
+stored in `method`, while the `results` section stores output from the workflow beyond observables
+of single configurations. For example, any ensemble-averaged quantity from a molecular dynamics
+simulation would be stored under `workflow/results`. Then, the `inputs`, `outputs`, and `tasks` sections
+are used to define the specifics of the workflow. For some [standard workflows](references/standard_workflows.md), e.g., geometry optimization and molecular
```
Comment (ndaelman-hu) [inline docs/writing_a_parser_plugin/computational.md]: A graph showing how `inputs`, `outputs`, and `tasks` are connected would be neat here.

Comment (ndaelman-hu) [inline docs/writing_a_parser_plugin/computational.md]: > Then, the `inputs`, `outputs`, and `tasks` sections ~~are used to~~ define the specifics of the workflow.

File: docs/writing_a_parser_plugin/computational.md [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+dynamics, the NOMAD [normalizers]() <!-- TODO Link to normalizer docs  --> will automatically populate these specifics. The parser must only create the
+appropriate workflow section. <!-- TODO Should give an example somewhere -->
```
Comment (ndaelman-hu) [inline docs/writing_a_parser_plugin/computational.md:56]: > The parser must only create the appropriate workflow section.

Pls specify which workflow sections have to be set by the parser: `workflow2` or these standard workflows.

File: docs/writing_a_parser_plugin/computational.md [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+
+## Recommended parser layout
+
+The following represents the recommended core structure for a computational parser, <!-- TODO general comp. or just atomistic?? -->
+typically implemented within `<parserproject>/parser.py`.
+
+### Imports
+
+The imports typically include the necessary generic python modules, the required MetaInfo
+classes from nomad, and additional nomad utilities, e.g., from `nomad.atomutils`.
```
Comment (ndaelman-hu) [inline docs/writing_a_parser_plugin/computational.md]: Not sure if this section is necessary: anyone who knows Python, knows about modules...
There is a certain amount of basic knowledge that they should have before attempting to program.

I do think that this could be an opportunity to point out that one always has to import the **schema from nomad**, as well as the **specific parser type**.

Also, I suggest diminishing the example: too much space for a simple point to make.

File: docs/writing_a_parser_plugin/computational.md [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+    sec_run = self.archive.m_create(Run)
+    sec_run.program = Program(name='<PARSERNAME>', version='unknown')
+
+    self.parse_method()
+    self.parse_system()
+    self.parse_calculation()
+    self.parse_workflow()
+```
+
+Then, the individual functions to populate that various MetaInfo sections can be defined:
```
Comment (ndaelman-hu) [inline docs/writing_a_parser_plugin/computational.md:153]: Yes, but this design is very problematic to maintain: it basically circumvents any type of scope.
I would not recommend people to add new parsers in this style.
Actually, the old parsers will have to be refactored, but this is a discussion to be had with Area D.

Instead, (backend) API-driven design is the way to go imho.

File: docs/writing_a_parser_plugin/computational.md [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+
+
+class <Parsername>Parser:
+    def __init__(self):
+        self.<mainfiletype>_parser = <Parsername><Mainfiletype>Parser()
+```
+
+The main class for your parser, `<Parsername>Parser`, will contain the bulk of the parsing routine,
+described further below. It may be useful to create a distinct class, `<Parsername><Mainfiletype>Parser`,
+for dealing with various filetypes that may be parsed throughout the entirety of the routine.
```
Comment (ndaelman-hu) [inline docs/writing_a_parser_plugin/computational.md:114]: I think that you could elaborate here on the distinction between mainfiles and other files.

File: docs/writing_a_parser_plugin/computational.md [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+However, in the simplest case of a single file type parsed, the entire parser can of course be
+implemented within a single class.
+
+In the following, we will walk through the layout of the `<Parsername>Parser` class. First,
+every parser class should have a "main" function called `parse()`, which will be called by NOMAD
+when the appropriate mainfile is found:
```
Comment (ndaelman-hu) [inline docs/writing_a_parser_plugin/computational.md:121]: The API is a bit more elaborate, including `init_parser()` which sets `mainfile`, `logger`, and `open`.
There are also functionalities to manipulate the stored data, but I would elaborate on those somewhere else. 
I'd also point the user to `nomad/parsing/parsers.py/run_parser` and `nomad/parsing/file_parser/file_parsers.py` for the full API.

File: docs/writing_a_parser_plugin/parser_plugin_overview.md [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+
+NOMAD uses parsers to convert raw data (for example, output from computational software, instruments,
+or electronic lab notebooks) into NOMAD's common Archive format. The following pages describe
+how to develop such a parser and integrate it within the NOMAD software. The goal is equip users
+with the required knowledge to contribute to and extend NOMAD.
+
+
+## Getting started
+
+In principle, it is possible to [develop a "local parser"](references/quick_parser_setup.md) that uses the *nomad-lab* package to parse raw data,
```
Comment (ndaelman-hu) [inline docs/writing_a_parser_plugin/parser_plugin_overview.md:12]: I would add a paragraph below explaining the benefits that the NOMAD package provides.
Both from the parsing perspective and the storage, there are some good reasons to use our package rather than starting from scratch.

**Review by JosePizarro3 — APPROVED** [review]
Same as with the H5MD: feel free to merge at will, and then we work on the content later on.

File: docs/writing_a_parser_plugin/parser_plugin_overview.md [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+without changing the NOMAD software itself. This allows a quick start for focusing on the parsing
+of the data itself, but is not relevant for full integration of your new parser into NOMAD.
```
Comment (ndaelman-hu) [inline docs/writing_a_parser_plugin/parser_plugin_overview.md:14]: > ~~without changing the NOMAD software itself.~~

I would say, "create your own plugins, e.g. for parsing."

> but is not relevant for full integration of your new parser into NOMAD.

Why did you add this disclaimer? I would simply remove this.

File: docs/writing_a_parser_plugin/parser_plugin_overview.md [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+### Parser organization
+
+The NOMAD parsers can be found within your local NOMAD git repo under
+`dependencies/parsers/`. The parsers are organized into the following individual projects
+(`dependencies/parsers/<parserproject>`) with their own corresponding repositories:
+
+* [atomistic](https://github.com/nomad-coe/atomistic-parsers) - Parsers for output from classical molecular simulations, e.g., from Gromacs, Lammps, etc.
+* [database](https://github.com/nomad-coe/database-parsers) - Parsers for various databases, e.g., OpenKim.
+* [eelsdb](https://github.com/nomad-coe/nomad-parser-eelsdb) - Parser for the EELS database (https://eelsdb.eu/; to be integrated in the database project).
+* [electronic](https://github.com/nomad-coe/electronic-parsers) - Parsers for output from electronic structure calculations, e.g., from Vasp, Fhiaims, etc.
```
Comment (ndaelman-hu) [inline docs/writing_a_parser_plugin/parser_plugin_overview.md]: ab initio instead of electronic structure?

File: docs/writing_a_parser_plugin/parser_plugin_overview.md [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+* [nexus](https://github.com/nomad-coe/nomad-parser-nexus) - Parsers for combining various instrument output formats and electronic lab notebooks.
+* [workflow](https://github.com/nomad-coe/workflow-parsers) - Parsers for output from codes that specialize in workflows, e.g., Aflow, FHI-vibes, QuantumEspressoPhonon, etc.
```
Comment (ndaelman-hu) [inline docs/writing_a_parser_plugin/parser_plugin_overview.md]: For workflows: **Parsers for output from task managers and workflow schedulers**

File: docs/writing_a_parser_plugin/parser_plugin_overview.md [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+`<parserproject>parser` or `<parserproject>parsers`, depending on if one or more
+parsers are contained within the project, respectively. In the case of multiple parsers, the files
+for individual parsers are contained within a corresponding subdirectory: `<parserproject>parsers/<parsername>`
+For example, the Quantum Espresso parser files are found in `dependencies/parsers/electronic/electronicparsers/quantumespresso/`.
+
+
+### Setting up your development branches
+
+We will first focus on the case of adding a new parser to an existing parser project.
+[Creating a new parser project](references/advanced_new_parser_project.md) will require a few extra steps.
```
Comment (ndaelman-hu) [inline docs/writing_a_parser_plugin/parser_plugin_overview.md:43]: Typical walkthroughs will show the few extra steps first. If they do not apply, the user will be told to skip them.

File: docs/writing_a_parser_plugin/parser_plugin_overview.md [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+The existing parser projects are stored within their own git repositories and then linked
+to the NOMAD software. All current parser projects are available at [nomad-coe](https://github.com/nomad-coe)
+(see also individual links above).
+
+You will first need to create new branches within *both* the NOMAD project and *also* within the corresponding
```
Comment (ndaelman-hu) [inline docs/writing_a_parser_plugin/parser_plugin_overview.md:50]: Also address getting an account

File: docs/writing_a_parser_plugin/parser_plugin_overview.md [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+Now we need to repeat this process for the parser project that we plan to extend. As above,
+create a new issue at the relevant parser project GitHub page. Using the identical
+issue title as you did for the NOMAD project above is ideal for clarity.
+On the page of the new issue, in the right sidebar under the subsection `Development`, click
+`Create a branch`. As above, the branch name should be automatically generated with the
+corresponding issue number and the title of the issue, and the default source branch should be `develop`
+(which can be seen by clicking `change source branch`).
+Under `What's next`, the default option should be `Checkout locally`, which is what we want in this case.
+Click `Create branch`, and then copy the provided commands to the clipboard, and run them within the parser
+project folder within your local NOMAD repo, i.e., `dependencies/parsers/<parserproject>`.
```
Comment (ndaelman-hu) [inline docs/writing_a_parser_plugin/parser_plugin_overview.md:79]: This all falls under basic git usage... Isn't there a git tutorial we could refer to?

File: docs/writing_a_parser_plugin/parser_shell.md [hunk 17]
```diff
@@ -0,0 +1,128 @@
+# Creating the shell for your parser
```
Comment (ndaelman-hu) [inline docs/writing_a_parser_plugin/parser_shell.md]: Be careful with the choice of words: **shell** is often used to refer to a terminal session (actually 1 layer deeper). I would use **Setting up the framework for your parser**

File: docs/writing_a_parser_plugin/parser_shell.md [hunk 18]
```diff
@@ -0,0 +1,128 @@
...
+
+In the following please note the naming conventions:
+
+- <parsername\> - flat case, e.g., quantumespresso
+
+- <ParserName\> - Pascal case, e.g., QuantumEspresso
+
+- <parser_name\> - snake case, e.g., quantum_espresso
+
+- <PARSERNAME\> - upper flat case, e.g., QUANTUMESPRESSO
```
Comment (ndaelman-hu) [inline docs/writing_a_parser_plugin/parser_shell.md:16]: Pls state which convention is used where, i.e. filenames, classes, etc.

File: docs/writing_a_parser_plugin/parser_tests.md [hunk 19]
```diff
@@ -0,0 +1,58 @@
+# Creating parser tests
+
+Each parser should have a series of associated (py)tests to ensure that future developments do
+not affect the intended parsing. These tests can be found under directly under  `<parserproject>/tests/`
+directory, labeled by the corresponding parser name: `test_parsername.py`.
+
+Here is an example template of such a `pytest` code:
```
Comment (ndaelman-hu) [inline docs/writing_a_parser_plugin/parser_tests.md:7]: The example below is way too big. I would instead specify some good practices (and maybe keep a small example):
- Test specific cases using representative calculation. We use real-world data here. Note that this data will be public under the aforementioned license, so consult the relevant parties.
- Use `assert`ions to test proper extraction. To facilitate lookup, refer to the data in its original units.
- Make the test objective clear by choosing appropriately sized functions and parametrization.

I would also link to a tutorial and reference on `pytest`.

File: docs/writing_a_parser_plugin/references/old_docs.md [hunk 20]
```diff
@@ -0,0 +1,338 @@
+# How to write a parser
```
Comment (ndaelman-hu) [inline docs/writing_a_parser_plugin/references/old_docs.md:1]: Is this copied from pre-existing documentation? I will be skipping it then.

File: docs/writing_a_parser_plugin/references/quick_installing_nomad_dev.md [hunk 21]
```diff
@@ -0,0 +1,24 @@
...
+You should be on the branch `origin/develop` by default.
+Create a virtual environment for your nomad installation:
+
+    pip install virtualenv
+    virtualenv -p `which python3` .pyenv
+    source .pyenv/bin/activate
+
+Now install nomad:
+
+    ./scripts/setup_dev_env.sh
```
Comment (ndaelman-hu) [inline docs/writing_a_parser_plugin/references/quick_installing_nomad_dev.md:24]: This is all in another documentation already

**Review by ndaelman-hu — COMMENTED** [review]
This is quite extensive already. I have mostly specific, topical feedback.
2 recurring points are:
- keep the examples smaller. Nobody wants to read a block of text. It's more important to point out the highlights. The code itself can do the rest.
- I would not repeat too much of what was already explained in the previous docs. Here we should just reference.

File: docs/writing_a_parser_plugin/computational.md (hunk 1 above)
Comment (JFRudzinski) [reply docs/writing_a_parser_plugin/computational.md]: @ndaelman-hu could you be a little more specific? I am not sure what you are asking for here.

File: docs/writing_a_parser_plugin/computational.md (hunk 4 above)
Comment (JFRudzinski) [reply docs/writing_a_parser_plugin/computational.md]: We might be able to just link to the workflow page here. I added a TODO for now.

Comment (JFRudzinski) [reply docs/writing_a_parser_plugin/computational.md]: done, thanks.

File: docs/writing_a_parser_plugin/computational.md (hunk 3 above)
Comment (JFRudzinski) [reply docs/writing_a_parser_plugin/computational.md:45]: I just made a TODO note about it for now. I don't really want to worry about changing everything right now and then possibly changing it back later.

File: docs/writing_a_parser_plugin/computational.md (hunk 5 above)
Comment (JFRudzinski) [reply docs/writing_a_parser_plugin/computational.md:56]: added TODO

File: docs/writing_a_parser_plugin/computational.md (hunk 6 above)
Comment (JFRudzinski) [reply docs/writing_a_parser_plugin/computational.md]: totally agree. This was just a place holder. I shortened it now.

File: docs/writing_a_parser_plugin/computational.md (hunk 7 above)
Comment (JFRudzinski) [reply docs/writing_a_parser_plugin/computational.md:153]: I agree that we should update this based on how we (Area C+D) think the parsers should be designed moving forward...but we have to all be on the same page here, so this is a discussion we should first have among Area C. I made a TODO for now to update it and then let's plan to discuss at some point in our coming meetings.

File: docs/writing_a_parser_plugin/computational.md (hunk 8 above)
Comment (JFRudzinski) [reply docs/writing_a_parser_plugin/computational.md:114]: Made a TODO

File: docs/writing_a_parser_plugin/parser_plugin_overview.md (hunk 10 above)
Comment (JFRudzinski) [reply docs/writing_a_parser_plugin/parser_plugin_overview.md:12]: I don't think this is the place for that. We need to be careful not to continually repeat explanations of things, or else the docs will be tedious to browse through. I think your suggested paragraph falls somewhere at the higher (i.e., broader) level of the docs. We can certainly link it though, and I will make a note.

File: docs/writing_a_parser_plugin/parser_plugin_overview.md (hunk 11 above)
Comment (JFRudzinski) [reply docs/writing_a_parser_plugin/parser_plugin_overview.md:14]: I am trying to make a distinction between the case that you want to implement new metainfo and the case that you are using the existing NOMAD metainfo. Maybe this is not necessary within your perspective of how a plugin should be designed, but it is for the current design of the parsers. I will just make a note that we should discuss this.

File: docs/writing_a_parser_plugin/parser_plugin_overview.md (hunk 12 above)
Comment (JFRudzinski) [reply docs/writing_a_parser_plugin/parser_plugin_overview.md]: @ndaelman-hu are you asking me or telling me? I cannot say myself. I was under the impression that some people (although I can't remember who) didn't like the term ab initio for some reason (perhaps it didn't fit some applications/calculations I don't know). Isn't electronic structure a broader/safer term? Or is it misleading for you?

File: docs/writing_a_parser_plugin/parser_plugin_overview.md (hunk 13 above)
Comment (JFRudzinski) [reply docs/writing_a_parser_plugin/parser_plugin_overview.md]: changed. thanks.

File: docs/writing_a_parser_plugin/parser_plugin_overview.md (hunk 15 above)
Comment (JFRudzinski) [reply docs/writing_a_parser_plugin/parser_plugin_overview.md:50]: I guess ultimately this will just link to a different section, I will make a TODO.

File: docs/writing_a_parser_plugin/parser_plugin_overview.md (hunk 16 above)
Comment (JFRudzinski) [reply docs/writing_a_parser_plugin/parser_plugin_overview.md:79]: @ndaelman-hu I agree with you in principle...however, there are many ways to use git, and working on a larger software project is not the same as using git for your personal repos or within a small group. The idea here is that we guide the to-be developer according to the standards of git usage within the NOMAD project. Moreover, these docs should be useful for not only providing guidance to external users, but also to train newcomers to the internal NOMAD team.

I do think probably we should move this to a reference (I will make a TODO), so it is not crowding the more important content, but I would like some sort of clear description of how the development process works within the NOMAD git project.

File: docs/writing_a_parser_plugin/parser_shell.md (hunk 17 above)
Comment (JFRudzinski) [reply docs/writing_a_parser_plugin/parser_shell.md]: changed.

File: docs/writing_a_parser_plugin/parser_shell.md (hunk 18 above)
Comment (JFRudzinski) [reply docs/writing_a_parser_plugin/parser_shell.md:16]: made a TODO

File: docs/writing_a_parser_plugin/references/old_docs.md (hunk 20 above)
Comment (JFRudzinski) [reply docs/writing_a_parser_plugin/references/old_docs.md:1]: Yes, this is just for our reference as we write the new docs, I will remove later.

File: docs/writing_a_parser_plugin/references/quick_installing_nomad_dev.md (hunk 21 above)
Comment (JFRudzinski) [reply docs/writing_a_parser_plugin/references/quick_installing_nomad_dev.md:24]: Yes, but I would like to keep a shortened version of this for users who maybe just need to remember the commands, but don't want to read through the remainder of the docs.

File: docs/writing_a_parser_plugin/parser_tests.md (hunk 19 above)
Comment (JFRudzinski) [reply docs/writing_a_parser_plugin/parser_tests.md:7]: agreed, I just copied your comment into a TODO for now

File: docs/writing_a_parser_plugin/parser_plugin_overview.md (hunk 14 above)
Comment (JFRudzinski) [reply docs/writing_a_parser_plugin/parser_plugin_overview.md:43]: I think we need to discuss this together at some later time, have already made a TODO while addressing one comment above.

File: docs/writing_a_parser_plugin/computational.md (hunk 2 above)
Comment (JFRudzinski) [reply docs/writing_a_parser_plugin/computational.md]: Not exactly sure what is meant here, just made a note now though and I will come back to it.

File: docs/writing_a_parser_plugin/computational.md (hunk 9 above)
Comment (JFRudzinski) [reply docs/writing_a_parser_plugin/computational.md:121]: made note of this comment, will return to it.

**Comment by JFRudzinski** [comment]
@ndaelman-hu thanks for all your comments. You will see that I followed up on some of them above, but for now I think I am just going to merge this initial version (since I don't want to work extensively on this now), and when I am ready to work on it I would probably like to meet with you briefly to discuss some of your comments. For now, I have made notes for all outstanding ones as TODOs within the docs.

File: docs/writing_a_parser_plugin/parser_plugin_overview.md (hunk 12 above)
Comment (JosePizarro3) [reply docs/writing_a_parser_plugin/parser_plugin_overview.md]: > Isn't electronic structure a broader/safer term?

It is, indeed. There are non-ab initio codes in electronic-parsers.
